src/java/main/org/apache/zookeeper/server/quorum/Leader.java

```diff
@@ -590,8 +590,8 @@ void lead() throws IOException, InterruptedException {
 
                     // check leader running status
                     if (!this.isRunning()) {
-                        shutdown("Unexpected internal error");
-                        return;
+                        shutdownMessage = "Unexpected internal error";
```

Comment:
Good catch @arshadmohammad. Yes, this is another code where it can leads to deadlock similar to ZK-2380. 

As we know, `shutdownMessage` is acting as flag. Could you please add comments to make this clear,
`// set shutdown flag`
`shutdownMessage` = "Unexpected internal error";`